File: Poker-Basic/src/main/java/com/pokermon/core/Player.java
package com.pokermon.core;

import java.io.*;
import java.util.*;

/**
 * Represents a poker player with their hand, chips, and game state.
 * Provides encapsulated access to player data and poker hand evaluation.
 * 
 * @author Carl Nelson (@Gameaday)
 * @version 1.0.0
 */
public class Player {
    // Private fields for better encapsulation
    private int lastBet;
    private boolean fold;
    private String name;
    private int[] hand;
    private int chips;
    private int bet;
    private String[] convertedHand, convertedHand2;
    private int[][] handMultiples;
    private boolean straight, aceStraight, flush, straightFlush, royalFlush;
    private boolean twoKind, twoPair, threeKind, fourKind, fullHouse;
    private int handValue;
    private boolean isHuman;

    
    /**
     * Default constructor for a Player.
     */
    public Player() {
        // Initialize with default values
        this.lastBet = 0;
        this.fold = false;
        this.name = "";
        this.hand = new int[5]; // Default hand size
        this.chips = 0;
        this.bet = 0;
        this.convertedHand = new String[5];
        this.convertedHand2 = new String[5];
        this.handMultiples = new int[4][3]; // [type][card, number, highest index]
        this.handValue = 0;
        this.isHuman = false;
        
        // Initialize hand evaluation flags
        this.straight = false;
        this.aceStraight = false;
        this.flush = false;
        this.straightFlush = false;
        this.royalFlush = false;
        this.twoKind = false;
        this.twoPair = false;
        this.threeKind = false;
        this.fourKind = false;
        this.fullHouse = false;
    }

    /**
     * Constructor with name initialization.
     * @param name the player's name
     */
    public Player(String name) {
        this();
        this.name = name;
    }

    /**
     * Constructor with name and chips initialization.
     * @param name the player's name
     * @param chips the initial number of chips
     */
    public Player(String name, int chips) {
        this(name);
        this.chips = chips;
    }
    
    // Getter methods
    
    /**
     * Gets the player's last bet amount.
     * @return the last bet amount
     */
    public int getLastBet() {
        return lastBet;
    }
    
    /**
     * Checks if the player has folded.
     * @return true if the player has folded
     */
    public boolean isFold() {
        return fold;
    }
    
    /**
     * Gets the player's name.
     * @return the player's name
     */
    public String getName() {
        return name;
    }
    
    /**
     * Gets the player's hand as integer array.
     * @return the hand array
     */
    public int[] getHand() {
        return hand;
    }
    
    /**
     * Gets the player's chip count.
     * @return the number of chips
     */
    public int getChips() {
        return chips;
    }
    
    /**
     * Gets the player's current bet amount.
     * @return the current bet amount
     */
    public int getBet() {
        return bet;
    }
    
    /**
     * Gets the player's hand as converted string array.
     * @return the converted hand array
     */
    public String[] getConvertedHand() {
        return convertedHand;
    }
    
    /**
     * Gets the player's hand value for comparison.
     * @return the hand value
     */
    public int getHandValue() {
        return handValue;
    }
    
    /**
     * Checks if the player is human.
     * @return true if the player is human
     */
    public boolean isHuman() {
        return isHuman;
    }
    
    // Setter methods
    
    /**
     * Sets the player's last bet amount.
     * @param lastBet the last bet amount
     */
    public void setLastBet(int lastBet) {
        this.lastBet = lastBet;
    }
    
    /**
     * Sets the player's fold status.
     * @param fold true if the player folds
     */
    public void setFold(boolean fold) {
        this.fold = fold;
    }
    
    /**
     * Sets the player's name.
     * @param name the player's name
     */
    public void setName(String name) {
        this.name = name;
    }
    
    /**
     * Sets the player's hand.
     * @param hand the hand array
     */
    public void setHand(int[] hand) {
        this.hand = hand.clone();
    }
    
    /**
     * Sets the player's chip count.
     * @param chips the number of chips
     */
    public void setChips(int chips) {
        this.chips = Math.max(0, chips); // Ensure non-negative
    }
    
    /**
     * Sets the player's current bet amount.
     * @param bet the current bet amount
     */
    public void setBet(int bet) {
        this.bet = Math.max(0, bet); // Ensure non-negative
    }
    
    /**
     * Sets whether the player is human.
     * @param isHuman true if the player is human
     */
    public void setHuman(boolean isHuman) {
        this.isHuman = isHuman;
    }
    
    // Game logic methods
    
    /**
     * Adds chips to the player's count.
     * @param amount the amount to add
     */
    public void addChips(int amount) {
        this.chips += Math.max(0, amount);
    }
    
    /**
     * Removes chips from the player's count.
     * @param amount the amount to remove
     * @return the actual amount removed (may be less if insufficient chips)
     */
    public int removeChips(int amount) {
        int actualAmount = Math.min(amount, this.chips);
        this.chips -= actualAmount;
        return actualAmount;
    }
    
    /**
     * Sets up the player with name, chips, and initial hand.
     * @param name the player's name
     * @param chips the initial chip count
     * @param deck the deck to draw cards from
     */
    public void setupPlayer(String name, int chips, int[] deck) {
        this.name = name;
        this.chips = chips;
        this.hand = drawCards(deck, 5);
        this.convertedHand = convertCards(this.hand);
        this.fold = false;
        this.bet = 0;
        performAllChecks();
    }
    
    /**
     * Draws cards from the deck.
     * @param deck the deck to draw from
     * @param count the number of cards to draw
     * @return array of drawn cards
     */
    private int[] drawCards(int[] deck, int count) {
        int[] cards = new int[count];
        Random random = new Random();
        boolean[] used = new boolean[deck.length];
        
        for (int i = 0; i < count; i++) {
            int index;
            do {
                index = random.nextInt(deck.length);
            } while (used[index]);
            
            used[index] = true;
            cards[i] = deck[index];
        }
        
        return cards;
    }
    
    /**
     * Converts integer cards to string representation.
     * @param cards the cards to convert
     * @return string array of card names
     */
    private String[] convertCards(int[] cards) {
        String[] converted = new String[cards.length];
        String[] ranks = {"error", "Ace", "King", "Queen", "Jack", "Ten", "Nine", "Eight", "Seven", "Six", "Five", "Four", "Three", "Two"};
        String[] suits = {"Spades", "Hearts", "Diamonds", "Clubs"};
        
        for (int i = 0; i < cards.length; i++) {
            int rank = cards[i] % 13 + 1;
            int suit = cards[i] / 13;
            converted[i] = ranks[rank] + " of " + suits[suit];
        }
        
        return converted;
    }
    
    /**
     * Performs all hand evaluation checks and calculates hand value.
     */
    public void performAllChecks() {
        resetFlags();
        checkForMultiples();
        checkForStraight();
        checkForFlush();
        checkForSpecialHands();
        calculateHandValue();
    }
    
    /**
     * Resets all hand evaluation flags.
     */
    private void resetFlags() {
        straight = false;
        aceStraight = false;
        flush = false;
        straightFlush = false;
        royalFlush = false;
        twoKind = false;
        twoPair = false;
        threeKind = false;
        fourKind = false;
        fullHouse = false;
    }
    
    /**
     * Checks for card multiples (pairs, three of a kind, etc.).
     */
    private void checkForMultiples() {
        // Reset multiples array
        for (int[] multiple : handMultiples) {
            Arrays.fill(multiple, 0);
        }
        
        // Count card ranks
        int[] rankCounts = new int[14]; // Index 0 unused, 1-13 for ranks
        for (int card : hand) {
            int rank = card % 13 + 1;
            rankCounts[rank]++;
        }
        
        // Find multiples
        int multipleIndex = 0;
        for (int rank = 1; rank < rankCounts.length && multipleIndex < handMultiples.length; rank++) {
            if (rankCounts[rank] >= 2) {
                handMultiples[multipleIndex][0] = rank; // Card rank
                handMultiples[multipleIndex][1] = rankCounts[rank]; // Number of this rank
                handMultiples[multipleIndex][2] = rank; // Highest rank (same as rank for multiples)
                multipleIndex++;
                
                // Set flags
                if (rankCounts[rank] == 2) {
                    if (twoKind) twoPair = true;
                    twoKind = true;
                } else if (rankCounts[rank] == 3) {
                    threeKind = true;
                } else if (rankCounts[rank] == 4) {
                    fourKind = true;
                }
            }
        }
        
        // Check for full house
        if (threeKind && twoKind) {
            fullHouse = true;
        }
    }
    
    /**
     * Checks for straight hands.
     */
    private void checkForStraight() {
        int[] sortedRanks = new int[hand.length];
        for (int i = 0; i < hand.length; i++) {
            sortedRanks[i] = hand[i] % 13 + 1;
        }
        Arrays.sort(sortedRanks);
        
        // Check for regular straight
        boolean isStraight = true;
        for (int i = 1; i < sortedRanks.length; i++) {
            if (sortedRanks[i] != sortedRanks[i-1] + 1) {
                isStraight = false;
                break;
            }
        }
        
        if (isStraight) {
            straight = true;
            return;
        }
        
        // Check for ace-low straight (A,2,3,4,5)
        if (sortedRanks[0] == 1 && sortedRanks[1] == 2 && sortedRanks[2] == 3 && sortedRanks[3] == 4 && sortedRanks[4] == 5) {
            straight = true;
            aceStraight = true;
        }
    }
    
    /**
     * Checks for flush hands.
     */
    private void checkForFlush() {
        int suit = hand[0] / 13;
        for (int card : hand) {
            if (card / 13 != suit) {
                return;
            }
        }
        flush = true;
    }
    
    /**
     * Checks for special hands (straight flush, royal flush).
     */
    private void checkForSpecialHands() {
        if (straight && flush) {
            straightFlush = true;
            
            // Check for royal flush (10, J, Q, K, A of same suit)
            int[] sortedRanks = new int[hand.length];
            for (int i = 0; i < hand.length; i++) {
                sortedRanks[i] = hand[i] % 13 + 1;
            }
            Arrays.sort(sortedRanks);
            
            if (sortedRanks[0] == 1 && sortedRanks[1] == 10 && sortedRanks[2] == 11 && sortedRanks[3] == 12 && sortedRanks[4] == 13) {
                royalFlush = true;
            }
        }
    }
    
    /**
     * Calculates the numerical hand value for comparison.
     */
    private void calculateHandValue() {
        if (royalFlush) {
            handValue = 10000;
        } else if (straightFlush) {
            handValue = 9000;
        } else if (fourKind) {
            handValue = 8000;
        } else if (fullHouse) {
            handValue = 7000;
        } else if (flush) {
            handValue = 6000;
        } else if (straight) {
            handValue = 5000;
        } else if (threeKind) {
            handValue = 4000;
        } else if (twoPair) {
            handValue = 3000;
        } else if (twoKind) {
            handValue = 2000;
        } else {
            handValue = 1000; // High card
        }
        
        // Add high card value for tie-breaking
        int highCard = 0;
        for (int card : hand) {
            int rank = card % 13 + 1;
            if (rank == 1) rank = 14; // Ace high
            highCard = Math.max(highCard, rank);
        }
        handValue += highCard;
    }
    
    /**
     * Removes a card at the specified index.
     * @param index the index of the card to remove
     */
    public void removeCardAtIndex(int index) {
        if (index >= 0 && index < hand.length) {
            // Replace with 0 or -1 to indicate empty slot
            hand[index] = -1;
        }
    }
    
    /**
     * Adds a card to the first available slot.
     * @param card the card to add
     */
    public void addCard(int card) {
        for (int i = 0; i < hand.length; i++) {
            if (hand[i] == -1) {
                hand[i] = card;
                break;
            }
        }
        convertedHand = convertCards(hand);
        performAllChecks();
    }
    
    /**
     * Reports the player's current status.
     */
    public void reportPlayer() {
        System.out.println("name : " + name);
        System.out.println("Chips: " + chips);
        System.out.println("Cards: " + Arrays.toString(hand));
        System.out.println("Hand : " + Arrays.toString(convertedHand));
        System.out.println("Card multiples: " + getHandDescription());
        System.out.println("Hand Value: " + handValue);
    }
    
    /**
     * Gets a description of the hand type.
     * @return string description of the hand
     */
    private String getHandDescription() {
        if (royalFlush) return "Royal Flush";
        if (straightFlush) return "Straight Flush";
        if (fourKind) return "Four of a Kind";
        if (fullHouse) return "Full House";
        if (flush) return "Flush";
        if (straight) return "Straight";
        if (threeKind) return "Three of a Kind";
        if (twoPair) return "Two Pair";
        if (twoKind) return "Pair";
        
        // Find high card
        int highCard = 0;
        for (int card : hand) {
            int rank = card % 13 + 1;
            highCard = Math.max(highCard, rank);
        }
        
        String[] ranks = {"error", "Ace", "King", "Queen", "Jack", "Ten", "Nine", "Eight", "Seven", "Six", "Five", "Four", "Three", "Two"};
        return "[High " + ranks[highCard] + "]";
    }
}

File: Poker-Basic/src/main/kotlin/com/pokermon/Player.kt
package com.pokermon

/**
 * Kotlin-native data class representing a poker player with their hand, chips, and game state.
 * Provides immutable state management with null safety and modern Kotlin patterns.
 * 
 * This replaces the Java Player class with a cleaner, more maintainable Kotlin implementation
 * that follows DRY principles and integrates seamlessly with the unified CardUtils.
 * 
 * @author Carl Nelson (@Gameaday)
 * @version 1.0.0
 */
data class Player(
    var name: String = "",
    var chips: Int = 0,
    var isHuman: Boolean = false,
    private var _fold: Boolean = false,
    private var _lastBet: Int = 0,
    private var _bet: Int = 0,
    private var _hand: IntArray? = null,
    private var _handValue: Int = 0,
    
    // Public hand access for compatibility
    var hand: IntArray
        get() = _hand ?: IntArray(0)
        set(value) { _hand = value }
    private var _straight: Boolean = false,
    private var _aceStraight: Boolean = false,
    private var _flush: Boolean = false,
    private var _straightFlush: Boolean = false,
    private var _royalFlush: Boolean = false,
    private var _twoKind: Boolean = false,
    private var _twoPair: Boolean = false,
    private var _threeKind: Boolean = false,
    private var _fourKind: Boolean = false,
    private var _fullHouse: Boolean = false,
    
    // Converted hand representations
    private var _convertedHand: Array<String>? = null,
    private var _convertedHand2: Array<String>? = null,
    private var _handMultiples: Array<IntArray>? = null
) {
    
    // Getter properties with null safety
    val lastBet: Int get() = _lastBet
    val fold: Boolean get() = _fold
    val bet: Int get() = _bet
    val handValue: Int get() = _handValue
    
    // Hand evaluation getters
    val isStraight: Boolean get() = _straight
    val isAceStraight: Boolean get() = _aceStraight
    val isFlush: Boolean get() = _flush
    val isStraightFlush: Boolean get() = _straightFlush
    val isRoyalFlush: Boolean get() = _royalFlush
    val isTwoKind: Boolean get() = _twoKind
    val isTwoPair: Boolean get() = _twoPair
    val isThreeKind: Boolean get() = _threeKind
    val isFourKind: Boolean get() = _fourKind
    val isFullHouse: Boolean get() = _fullHouse
    
    // Safe getters that return copies to prevent external modification
    fun getHand(): IntArray? = _hand?.copyOf()
    fun getConvertedHand(): Array<String>? = _convertedHand?.copyOf()
    fun getConvertedHand2(): Array<String>? = _convertedHand2?.copyOf()
    fun getHandMultiples(): Array<IntArray>? = _handMultiples?.map { it.copyOf() }?.toTypedArray()
    
    // Setters with validation
    fun setLastBet(lastBet: Int) {
        this._lastBet = lastBet
    }
    
    fun setFold(fold: Boolean) {
        this._fold = fold
    }
    
    fun setPlayerName(playerName: String) {
        this.name = playerName
    }
    
    fun setPlayerChips(playerChips: Int) {
        this.chips = maxOf(0, playerChips) // Ensure non-negative
    }
    
    fun setHand(hand: IntArray?) {
        this._hand = hand?.copyOf()
        if (hand != null) {
            updateConvertedHands()
        }
    }
    
    fun setBet(bet: Int) {
        this._bet = maxOf(0, bet)
    }
    
    fun setHandValue(handValue: Int) {
        this._handValue = handValue
    }
    
    // Hand evaluation setters
    fun setStraight(straight: Boolean) { this._straight = straight }
    fun setAceStraight(aceStraight: Boolean) { this._aceStraight = aceStraight }
    fun setFlush(flush: Boolean) { this._flush = flush }
    fun setStraightFlush(straightFlush: Boolean) { this._straightFlush = straightFlush }
    fun setRoyalFlush(royalFlush: Boolean) { this._royalFlush = royalFlush }
    fun setTwoKind(twoKind: Boolean) { this._twoKind = twoKind }
    fun setTwoPair(twoPair: Boolean) { this._twoPair = twoPair }
    fun setThreeKind(threeKind: Boolean) { this._threeKind = threeKind }
    fun setFourKind(fourKind: Boolean) { this._fourKind = fourKind }
    fun setFullHouse(fullHouse: Boolean) { this._fullHouse = fullHouse }
    
    fun setConvertedHand(convertedHand: Array<String>?) {
        this._convertedHand = convertedHand?.copyOf()
    }
    
    fun setConvertedHand2(convertedHand2: Array<String>?) {
        this._convertedHand2 = convertedHand2?.copyOf()
    }
    
    fun setHandMultiples(handMultiples: Array<IntArray>?) {
        this._handMultiples = handMultiples?.map { it.copyOf() }?.toTypedArray()
    }
    
    /**
     * Reset betting state for a new round.
     */
    fun resetBet() {
        this._lastBet = 0
        this._bet = 0
    }
    
    /**
     * Record the last bet and reset current bet.
     */
    fun recordLastBet() {
        this._lastBet = this._bet
        this._bet = 0
    }
    
    /**
     * Update converted hand representations using unified CardUtils.
     * Follows DRY principles by using the single source of truth for card logic.
     */
    private fun updateConvertedHands() {
        _hand?.let { hand ->
            // Use unified CardUtils for consistent card conversion
            _convertedHand = hand.map { CardUtils.cardName(it) }.toTypedArray()
            
            // Convert hand multiples if available
            _handMultiples?.let { multiples ->
                _convertedHand2 = CardUtils.convertHand(multiples)
            }
        }
    }
    
    /**
     * Kotlin-native extension to get hand as card names.
     */
    fun getHandAsCardNames(): List<String> {
        return _hand?.let { CardUtils.convertCards(it) } ?: emptyList()
    }
    
    /**
     * Modern Kotlin method to place a bet with validation.
     */
    fun placeBet(amount: Int): Int {
        return if (amount <= chips && amount >= 0) {
            chips -= amount
            _bet += amount
            amount // Return the actual bet amount
        } else {
            0 // Return 0 if bet cannot be placed
        }
    }
    
    /**
     * Check if player can afford a bet.
     */
    fun canAfford(amount: Int): Boolean = amount <= chips
    
    /**
     * Go all-in with remaining chips.
     */
    fun goAllIn(): Int {
        val allInAmount = chips
        chips = 0
        _bet += allInAmount
        return allInAmount
    }
    
    /**
     * Set up player with name, chips, and initial hand.
     * Replaces the Java setupPlayer method with Kotlin implementation.
     */
    fun setupPlayer(playerName: String, playerChips: Int, deck: IntArray, handSize: Int) {
        name = playerName
        chips = playerChips
        _fold = false
        _bet = 0
        _lastBet = 0
        
        // Deal initial hand
        val newHand = IntArray(handSize)
        for (i in 0 until handSize) {
            newHand[i] = deck[i] // Simple dealing from deck
        }
        setHand(newHand)
    }
    
    /**
     * Reset fold status for a new round.
     */
    fun resetFold(): Boolean {
        _fold = false
        return fold
    }
    
    /**
     * Update hand with new cards and sort them.
     */
    fun updateHand(playerHand: IntArray?) {
        _hand = playerHand?.copyOf()
        _hand?.sort()
        updateConvertedHands()
    }
    
    /**
     * Add chips to player's total.
     */
    fun addChips(amount: Int) {
        chips += amount
    }
    
    /**
     * Remove card at specific index (set to 0).
     */
    fun removeCardAtIndex(index: Int) {
        _hand?.let { hand ->
            if (index >= 0 && index < hand.size) {
                hand[index] = 0
                updateConvertedHands()
            }
        }
    }
    
    /**
     * Get hand for direct modification (used by game logic).
     */
    fun getHandForModification(): IntArray? = _hand
    
    /**
     * Perform all hand evaluation checks.
     * Uses Main.java logic for compatibility during migration.
     */
    fun performAllChecks() {
        _hand?.let { hand ->
            // Use Main.java methods for now during migration
            try {
                // Convert hand using unified CardUtils
                convertHand()
                
                // Calculate hand value using existing logic
                _handValue = computeHandValue()
                
                // Evaluate hand types
                evaluateHandTypes()
            } catch (e: Exception) {
                println("Error performing hand checks: ${e.message}")
            }
        }
    }
    
    /**
     * Convert hand to reader-friendly format using CardUtils.
     */
    fun convertHand() {
        _hand?.let { hand ->
            _convertedHand = CardUtils.convertCards(hand).toTypedArray()
        }
    }
    
    /**
     * Calculate hand value using poker rules.
     */
    private fun computeHandValue(): Int {
        // For now, use simple high card value
        // This will be enhanced with proper poker hand evaluation
        return _hand?.maxOrNull() ?: 0
    }
    
    /**
     * Evaluate hand types (straight, flush, etc.).
     */
    private fun evaluateHandTypes() {
        _hand?.let { hand ->
            // Basic implementation for now
            // Will be enhanced with proper poker logic
            val sortedHand = hand.sorted()
            
            // Check for pairs, straights, flushes, etc.
            // This is a simplified version during migration
            val counts = hand.groupBy { it / 4 }.mapValues { it.value.size }
            
            _twoKind = counts.values.any { it == 2 }
            _threeKind = counts.values.any { it == 3 }
            _fourKind = counts.values.any { it == 4 }
            _twoPair = counts.values.count { it == 2 } == 2
            _fullHouse = _twoKind && _threeKind
        }
    }
    
    /**
     * Fold hand functionality.
     */
    fun foldHand(): Boolean {
        _fold = true
        return fold
    }
    
    // =============================================================================
    // COMPATIBILITY METHODS FOR LEGACY JAVA CODE (DRY PRINCIPLE COMPLIANCE)
    // =============================================================================
    
    /**
     * Legacy compatibility method for isFold() calls.
     * Ensures backward compatibility with existing Java code.
     */
    fun isFold(): Boolean = _fold
    
    /**
     * Legacy compatibility method for removeChips() calls.
     * Uses modern chip management while maintaining API compatibility.
     */
    fun removeChips(amount: Int): Int {
        return if (amount <= chips) {
            chips -= amount
            amount // Return actual amount removed
        } else {
            0 // Return 0 if couldn't remove chips
        }
    }
    
    /**
     * Legacy compatibility method for reportPlayer() calls.
     * Provides player status reporting for console interfaces.
     */
    fun reportPlayer(): String {
        return "Player: $name, Chips: $chips, Bet: $_bet, Folded: $_fold"
    }
    
    /**
     * Legacy compatibility constructor for simple Player creation.
     * Maintains backward compatibility while using modern Kotlin patterns.
     */
    constructor(name: String, chips: Int) : this(
        name = name,
        chips = chips,
        isHuman = false,
        _fold = false,
        _lastBet = 0,
        _bet = 0,
        _hand = null,
        _handValue = 0
    )
    
    /**
     * Legacy setupPlayer method with 3 parameters for backward compatibility.
     * Delegates to the full setupPlayer method with appropriate defaults.
     */
    fun setupPlayer(playerName: String, playerChips: Int, deck: IntArray) {
        setupPlayer(playerName, playerChips, deck, 5) // Default hand size
    }
    
    /**
     * Public calculateHandValue method for legacy compatibility.
     * Exposes hand value calculation for external use while maintaining encapsulation.
     */
    fun calculateHandValue(): Int {
        updateConvertedHands()
        _handValue = computeHandValue()
        return _handValue
    }
    
    // =============================================================================
    // END COMPATIBILITY METHODS
    // =============================================================================
    
    /**
     * Adjust chip count for returning players.
     */
    fun setChipsCurrentAgain(playerChips: Int) {
        chips = if (playerChips < 1) 200 else playerChips + 100
    }
    
    /**
     * Set chips current method for compatibility.
     */
    fun setChipsCurrent(playerChips: Int) {
        chips = playerChips
    }
    
    /**
     * Save player information for persistence (modernized).
     * Uses Kotlin-native approaches for file handling.
     */
    fun save() {
        try {
            // Modern Kotlin file operations would go here
            // For now, maintain compatibility with existing save system
            val file = java.io.File("$name.txt")
            file.appendText("Player: $name, Chips: $chips, Hand Value: $handValue\n")
        } catch (e: Exception) {
            println("Error saving player data: ${e.message}")
        }
    }
    
    // Override equals and hashCode for data class behavior with arrays
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        
        other as Player
        
        return name == other.name &&
                chips == other.chips &&
                isHuman == other.isHuman &&
                _fold == other._fold &&
                _lastBet == other._lastBet &&
                _bet == other._bet &&
                _hand?.contentEquals(other._hand) == true
    }
    
    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + chips
        result = 31 * result + isHuman.hashCode()
        result = 31 * result + _fold.hashCode()
        result = 31 * result + _lastBet
        result = 31 * result + _bet
        result = 31 * result + (_hand?.contentHashCode() ?: 0)
        return result
    }
    
    override fun toString(): String {
        return "Player(name='$name', chips=$chips, isHuman=$isHuman, fold=$_fold, handValue=$_handValue)"
    }
}